Geometry_Engine/Query/GeometryHash.cs

```diff
@@ -0,0 +1,446 @@
+            "The ISubParts() methods is able to return the 'primitive' curves that a curve is composed of. " +
+            "The GeometryHashes are then calculated for the individual parts and concatenated.")]
+        private static double[] GeometryHash(this ICurve curve, double translationFactor)
+        {
+            List<ICurve> subParts = curve.ISubParts().ToList();
+
+            List<double> hashes = new List<double>();
+
+            //Add hash ignoring endpoint for all but last curve
+            for (int i = 0; i < subParts.Count - 1; i++)
+            {
+                hashes.AddRange(GeometryHash(subParts[i] as dynamic, translationFactor, true));
+            }
+            //Include endpoint for hasing for last curve
+            hashes.AddRange(GeometryHash(subParts.Last() as dynamic, translationFactor, false));
+
+            return hashes.ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for an Arc is calculated as the GeometryHash of the start, end and middle point of the Arc.")]
+        private static double[] GeometryHash(this Arc curve, double translationFactor, bool skipEndPoint= false)
+        {
+            translationFactor += (int)TypeTranslationFactor.Arc;
+
+            IEnumerable<double> hash = curve.StartPoint().ToDoubleArray(translationFactor)
+               .Concat(curve.PointAtParameter(0.5).ToDoubleArray(translationFactor));
+
+            if (!skipEndPoint)
+                hash = hash.Concat(curve.EndPoint().ToDoubleArray(translationFactor));
+
+            return hash.ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for an Circle is calculated as the GeometryHash of the start, 1/3rd and 2/3rd points of the Circle.")]
+        private static double[] GeometryHash(this Circle curve, double translationFactor, bool skipEndPoint = false)
+        {
+            // The input `skipEndPoint` is not used here because Ellipses cannot be part of Polycurves.
+
+            translationFactor += (int)TypeTranslationFactor.Circle;
+
+            IEnumerable<double> hash = curve.StartPoint().ToDoubleArray(translationFactor)
+                   .Concat(curve.PointAtParameter(0.33).ToDoubleArray(translationFactor))
+                   .Concat(curve.PointAtParameter(0.66).ToDoubleArray(translationFactor));
+
+            return hash.ToArray();
+        }
+
+        /***************************************************/
+
+        [Description("The GeometryHash for an Ellipse is calculated as the GeometryHash of the start, 1/3rd and 2/3rd points of the Ellipse.")]
+        private static double[] GeometryHash(this Ellipse curve, double translationFactor, bool skipEndPoint = false)
+        {
+            // The input `skipEndPoint` is not used here because Ellipses cannot be part of Polycurves.
```

Comment:
As above
```suggestion
            // The input `skipEndPoint` is not used here because Ellipses do not have a clearly defined endpoint to be used in a chain of segment curves.
```